packages/swapper/src/swappers/RelaySwapper/utils/getTrade.ts

```diff
@@ -0,0 +1,365 @@
+  } = input
+
+  const slippageToleranceBps = _slippageTolerancePercentageDecimal
+    ? convertDecimalPercentageToBasisPoints(_slippageTolerancePercentageDecimal).toFixed()
+    : undefined
+
+  const sellRelayChainId = relayChainMap[sellAsset.chainId]
+  const buyRelayChainId = relayChainMap[buyAsset.chainId]
+
+  if (sellRelayChainId === undefined) {
+    return Err(
+      makeSwapErrorRight({
+        message: `asset '${sellAsset.name}' on chainId '${sellAsset.chainId}' not supported`,
+        code: TradeQuoteError.UnsupportedTradePair,
+      }),
+    )
+  }
+
+  if (buyRelayChainId === undefined) {
+    return Err(
+      makeSwapErrorRight({
+        message: `asset '${buyAsset.name}' on chainId '${buyAsset.chainId}' not supported`,
+        code: TradeQuoteError.UnsupportedTradePair,
+      }),
+    )
+  }
+
+  const sendAddress = (() => {
+    if (input.quoteOrRate === 'rate') {
+      if (input.sendAddress) return input.sendAddress
+
+      // @TODO: Support solana and BTC addresses according to relay implementation when
+      // wallet is not connected
+      return DEFAULT_RELAY_EVM_USER_ADDRESS
+    }
+
+    return input.sendAddress
+  })()
+
+  const recipient = (() => {
+    if (input.quoteOrRate === 'rate') {
+      if (input.receiveAddress) return input.receiveAddress
+
+      // @TODO: Support solana and BTC addresses according to relay implementation when
+      // wallet is not connected
+      return DEFAULT_RELAY_EVM_USER_ADDRESS
+    }
+
+    return input.receiveAddress
+  })()
+
+  const maybeQuote = await fetchRelayTrade(
+    {
+      originChainId: sellRelayChainId,
+      originCurrency: getRelayEvmAssetAddress(sellAsset),
+      destinationChainId: buyRelayChainId,
+      destinationCurrency: getRelayEvmAssetAddress(buyAsset),
+      tradeType: 'EXACT_INPUT',
+      amount: sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      recipient,
+      user: sendAddress,
+      slippageTolerance: slippageToleranceBps,
+      refundOnOrigin: true,
+    },
+    deps.config,
+  )
+
+  if (maybeQuote.isErr()) return Err(maybeQuote.unwrapErr())
+
+  const { data: quote } = maybeQuote.unwrap()
+
+  if (!quote.details) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Relay quote details not found',
+      }),
+    )
+  }
+
+  const { slippageTolerance, rate, currencyOut, timeEstimate } = quote.details
+
+  if (!slippageTolerance) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Relay quote slippage tolerance not found',
+      }),
+    )
+  }
+
+  if (!rate) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Relay quote rate not found',
+      }),
+    )
+  }
+
+  if (!currencyOut) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Relay quote relayToken not found',
+      }),
+    )
+  }
+
+  if (!quote.fees?.gas) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Relay quote gas fees not found',
+      }),
+    )
+  }
```

Comment:
Seems a lot but everything on relay type side is optional so... we have to make it safe

Reply:
See https://github.com/shapeshift/web/pull/9188#discussion_r2030281471

Reply:
In [29fc44d](https://github.com/shapeshift/web/pull/9188/commits/29fc44d8ce158086f2cbec2c3c7ee0dd44302dac)